Memory usage regression

I am noticing an important regression in GC heap sizes and working set measurement on ASP.NET benchmarks (most of them).

These are the charts for Json MVC.

![image](https://user-images.githubusercontent.com/1165805/182445204-f087594a-7ae6-4e13-82ba-56c0bcc9866e.png)

![image](https://user-images.githubusercontent.com/1165805/182445121-2bf4dabe-415c-4665-bc58-8ae3e4bb9092.png)

~~An early range of changes points to these commits: https://github.com/dotnet/runtime/compare/cfe26ec...68bd58e~~
Smallest set of changes reproducing the regression: https://github.com/dotnet/runtime/compare/cfe26ec...e415018

One of these changes impacted some GC work: https://github.com/dotnet/runtime/pull/72923

do you know why I'm getting this? 
```
C:\temp\mvc70>dotnet new mvc
The template "ASP.NET Core Web App (Model-View-Controller)" was created successfully.
This template contains technologies from parties other than Microsoft, see https://aka.ms/aspnetcore/7.0-third-
party-notices for details.

Processing post-creation actions...
Restoring C:\temp\mvc70\mvc70.csproj:
  Determining projects to restore...
  Restored C:\temp\mvc70\mvc70.csproj (in 83 ms).
Restore succeeded.

C:\temp\mvc70>dotnet publish -c Release --self-contained -r win-x64
MSBuild version 17.4.0-preview-22368-02+c8492483a for .NET
  Determining projects to restore...
C:\temp\mvc70\mvc70.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Runtime.win-x64 with v
ersion (= 7.0.0-preview.7.22375.6)
C:\temp\mvc70\mvc70.csproj : error NU1102:   - Found 92 version(s) in nuget.org [ Nearest version: 7.0.0-previ
ew.6.22324.4 ]
C:\temp\mvc70\mvc70.csproj : error NU1102:   - Found 0 version(s) in Microsoft Visual Studio Offline Packages
C:\temp\mvc70\mvc70.csproj : error NU1102: Unable to find package Microsoft.WindowsDesktop.App.Runtime.win-x64
 with version (= 7.0.0-preview.7.22377.1)
C:\temp\mvc70\mvc70.csproj : error NU1102:   - Found 90 version(s) in nuget.org [ Nearest version: 7.0.0-previ
ew.6.22351.3 ]
C:\temp\mvc70\mvc70.csproj : error NU1102:   - Found 0 version(s) in Microsoft Visual Studio Offline Packages
C:\temp\mvc70\mvc70.csproj : error NU1102: Unable to find package Microsoft.AspNetCore.App.Runtime.win-x64 wit
h version (= 7.0.0-preview.7.22376.6)
C:\temp\mvc70\mvc70.csproj : error NU1102:   - Found 93 version(s) in nuget.org [ Nearest version: 7.0.0-previ
ew.6.22330.3 ]
C:\temp\mvc70\mvc70.csproj : error NU1102:   - Found 0 version(s) in Microsoft Visual Studio Offline Packages
  Failed to restore C:\temp\mvc70\mvc70.csproj (in 297 ms).
```

--- a/src/coreclr/gc/gc.cpp
+++ b/src/coreclr/gc/gc.cpp
@@ -28421,6 +28421,7 @@ void gc_heap::plan_phase (int condemned_gen_number)
             heap_segment* sseg = seg;
             do
             {
+                heap_segment_saved_allocated(seg) = 0;
                 uint8_t* start_unmarked = heap_segment_mem (seg);
 #ifndef USE_REGIONS
                 // shorten it to minimum

--- a/src/coreclr/gc/gc.cpp
+++ b/src/coreclr/gc/gc.cpp
@@ -28245,9 +28245,9 @@ void gc_heap::add_plug_in_condemned_info (generation* gen, size_t plug_size)
 
 inline void save_allocated(heap_segment* seg)
 {
-#ifndef MULTIPLE_HEAP
+#ifndef MULTIPLE_HEAPS
     if (!heap_segment_saved_allocated(seg))
-#endif // !MULTIPLE_HEAP
+#endif // !MULTIPLE_HEAPS
     {
         heap_segment_saved_allocated (seg) = heap_segment_allocated (seg);
     }

